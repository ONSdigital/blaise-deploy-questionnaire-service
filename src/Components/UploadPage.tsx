import React, {ReactElement, useState} from "react";
import ONSErrorPanel from "./ONSDesignSystem/ONSErrorPanel";
import {ONSPanel} from "./ONSDesignSystem/ONSPanel";
import {Link, Redirect} from "react-router-dom";
import {ONSUpload} from "./ONSDesignSystem/ONSUpload";
import {ONSButton} from "./ONSDesignSystem/ONSButton";
import uploadImage from "../Upload";

interface Props {
    external_client_url: string
}


function UploadPage(props: Props): ReactElement {
    const [redirect, setRedirect] = useState<boolean>(false);
    const [file, setFile] = useState<FileList>();

    function UploadFile() {
        console.log("Yeah that didn't work");
    }

    const handleFileChange = (selectorFiles: FileList | null) => {
        console.log(selectorFiles);
        if (selectorFiles !== null) {
            setFile(selectorFiles);
        }
    };

    return (
        <>
            {
                redirect && <Redirect to="/UploadSummary"/>
            }
            <Link to="/">
                Previous
            </Link>
            <h1>
                Deploy a questionnaire file
            </h1>
            <ONSPanel>
                <p>
                    When a questionnaire file is selected and you continue to deploy this questionnaire file, <b>this
                    may take a few minutes</b>.
                    <br/>
                    <br/>
                    Given this, <b>do not navigate away</b> from this page during this process. You will be re-directed
                    when there is an update regarding the deploy of the questionnaire.
                </p>
            </ONSPanel>

            <ONSUpload label="Select survey package" description="File type accepted is .bpkg only" fileName="Package"
                       fileID="ID" accept="bpkg" onChange={(e) => handleFileChange(e.target.files)}/>
            <ONSButton label="Continue" primary={true} onClick={() => UploadFile()}/>
        </>
    );
}

export default UploadPage;
