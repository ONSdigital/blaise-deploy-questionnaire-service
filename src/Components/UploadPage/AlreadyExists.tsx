import React, {useState} from "react";
import {useHistory} from "react-router-dom";
import {ONSButton} from "../ONSDesignSystem/ONSButton";

interface Props {
    instrumentName: string
    ConfirmInstrumentOverride: any
    loading: boolean
}

function AlreadyExists({instrumentName, ConfirmInstrumentOverride, loading} : Props) {
    const [confirm, setConfirm] = useState<boolean>(false);
    const history = useHistory();

    function confirmOption() {

        if (!confirm) {
            history.push("/");
            return;
        }
        ConfirmInstrumentOverride();
    }

    return (
        <>
            <h1>Questionnaire file <em className="highlight">{instrumentName}</em> already exists in the system.
            <br/>
            What action do you want to take?
            </h1>

            <form onSubmit={() => confirmOption()}>
                <fieldset className="fieldset">
                    <legend className="fieldset__legend">
                    </legend>
                    <div className="radios__items">
                        <p className="radios__item">
                        <span className="radio">
                        <input
                            type="radio"
                            id="cancel-keep"
                            className="radio__input js-radio "
                            value="False"
                            name="confirm-delete"
                            aria-label="Yes"
                            onChange={() => setConfirm(false)}
                        />
                        <label className="radio__label " htmlFor="cancel-keep">
                            Cancel and keep original questionnaire
                        </label>
                    </span>
                        </p>
                        <br/>
                        <p className="radios__item">
                        <span className="radio">
                        <input
                            type="radio"
                            id="confirm-overwrite"
                            className="radio__input js-radio "
                            value="True"
                            name="confirm-delete"
                            aria-label="No"
                            onChange={() => setConfirm(true)}
                        />
                        <label className="radio__label " htmlFor="confirm-overwrite">
                            Overwrite the entire questionnaire
                        </label>
                    </span></p>
                    </div>
                </fieldset>

                <br/>
                <ONSButton
                    label={"Save"}
                    primary={true}
                    id="confirm-save"
                    loading={loading}
                    onClick={() => confirmOption()}/>
            </form>
        </>
    );
}

export default AlreadyExists;
