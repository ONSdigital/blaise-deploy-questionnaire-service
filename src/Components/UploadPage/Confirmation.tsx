import React, {useState} from "react";
import {useHistory} from "react-router-dom";
import {ONSButton} from "../ONSDesignSystem/ONSButton";

interface Props {
    instrumentName: string
    UploadFile: any
    loading: boolean
}

function Confirmation({instrumentName, UploadFile, loading}: Props) {
    const [confirm, setConfirm] = useState<boolean>(false);
    const [message, setMessage] = useState<string>("");
    const history = useHistory();

    function confirmOption() {

        if (!confirm) {
            history.push("/");
            return;
        }
        UploadFile();
    }

    return (
        <>
            <h1>
                Are you sure you want to overwrite the entire questionnaire <em
                className="highlight">{instrumentName}</em>?
            </h1>

            <p>
                {message}
            </p>

            <div className="panel panel--warn panel--no-title">
                <span className="panel__icon" aria-hidden="true">!</span>
                <div className="panel__body">
                    All existing questionnaire information will be deleted
                </div>
            </div>

            <form onSubmit={() => confirmOption()}>
                <fieldset className="fieldset">
                    <legend className="fieldset__legend">
                    </legend>
                    <div className="radios__items">

                        <p className="radios__item">
                        <span className="radio">
                        <input
                            type="radio"
                            id="confirm-overwrite"
                            className="radio__input js-radio "
                            value="True"
                            name="confirm-delete"
                            aria-label="No"
                            onChange={() => setConfirm(true)}
                        />
                        <label className="radio__label " htmlFor="confirm-overwrite">
                            Yes, overwrite questionnaire
                        </label>
                    </span></p>
                           <br/>
                        <p className="radios__item">
                        <span className="radio">
                        <input
                            type="radio"
                            id="cancel-keep"
                            className="radio__input js-radio "
                            value="False"
                            name="confirm-delete"
                            aria-label="Yes"
                            onChange={() => setConfirm(false)}
                        />
                        <label className="radio__label " htmlFor="cancel-keep">
                            No, do not overwrite questionnaire
                        </label>
                    </span>
                        </p>
                    </div>
                </fieldset>

                <br/>
                <ONSButton
                    label={"Continue"}
                    primary={true}
                    loading={loading}
                    onClick={() => confirmOption()}/>
                {!loading &&
                <ONSButton
                    label={"Cancel"}
                    primary={false}
                    onClick={() => confirmOption()}/>
                }
            </form>
        </>
    );
}

export default Confirmation;
