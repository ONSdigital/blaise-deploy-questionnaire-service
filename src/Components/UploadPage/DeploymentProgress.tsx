import React, {ReactElement} from "react";
import styled from "styled-components";

interface Props {
    instrumentName: string
    isVerifyIsInstalled: string
    isInstalling: string
    isUploading: string
    isVerifyingUpload: string
    isInitialisingUpload: string
    uploadPercentage: number
}

export const step_status = {
    NOT_STARTED: "Not Started",
    IN_PROGRESS: "In Progress",
    COMPLETE: "Completed",
    FAILED: "Failed",
};


const SpaciousTableRow = styled.tr`
  td:first-of-type {
    padding: 1.3rem 0 1.3rem;
  }

  td {
    width: 1rem;
    padding: 1.3rem 0 1.3rem 1rem;
  }
`;

const StyledSVG = styled.svg`
  margin-right: 5px
`;

function DeploymentSummary({
                               instrumentName,
                               isVerifyIsInstalled,
                               isInstalling,
                               isUploading,
                               uploadPercentage,
                               isVerifyingUpload,
                               isInitialisingUpload
                           }: Props): ReactElement {

    function getIsVerifyIsInstalled() {
        switch (isVerifyIsInstalled) {
            case step_status.NOT_STARTED:
                return <span className="status status--pending">Not started</span>;
            case step_status.IN_PROGRESS:
                return <><LoadingIcon/>Verifying</>;
            case step_status.COMPLETE:
                return <span className="status status--success">Completed</span>;
            case step_status.FAILED:
                return <span className="status status--error">Failed</span>;
        }
        return <p>hello</p>;
    }

    function getIsInitialisingUpload() {
        switch (isInitialisingUpload) {
            case step_status.NOT_STARTED:
                return <span className="status status--pending">Not started</span>;
            case step_status.IN_PROGRESS:
                return <><LoadingIcon/>Initialing</>;
            case step_status.COMPLETE:
                return <span className="status status--success">Completed</span>;
            case step_status.FAILED:
                return <span className="status status--error">Failed</span>;
        }
        return <p>hello</p>;
    }

    function getIsUploading() {
        switch (isUploading) {
            case step_status.NOT_STARTED:
                return <span className="status status--pending">Not started</span>;
            case step_status.IN_PROGRESS:
                return <><LoadingIcon/>Uploading {uploadPercentage}%</>;
            case step_status.COMPLETE:
                return <span className="status status--success">Completed</span>;
            case step_status.FAILED:
                return <span className="status status--error">Failed</span>;
        }
        return <p>hello</p>;
    }

    function getIsVerifyingUpload() {
        switch (isVerifyingUpload) {
            case step_status.NOT_STARTED:
                return <span className="status status--pending">Not started</span>;
            case step_status.IN_PROGRESS:
                return <><LoadingIcon/>Verifying</>;
            case step_status.COMPLETE:
                return <span className="status status--success">Completed</span>;
            case step_status.FAILED:
                return <span className="status status--error">Failed</span>;
        }
        return <p>hello</p>;
    }

    function getIsInstalling() {
        switch (isInstalling) {
            case step_status.NOT_STARTED:
                return <span className="status status--pending">Not started</span>;
            case step_status.IN_PROGRESS:
                return <><LoadingIcon/>Installing</>;
            case step_status.COMPLETE:
                return <span className="status status--success">Completed</span>;
            case step_status.FAILED:
                return <span className="status status--error">Failed</span>;
        }
        return <p>hello</p>;
    }

    return (
        <>


            <table id="deployment-progress-table" className="table summary__items  u-mt-m">
                <thead className="table__head">
                <tr className="table__row">
                    <th scope="col" className="table__header ">
                        <span>Deployment step</span>
                    </th>
                    <th scope="col" className="table__header ">
                        <span>Status</span>
                    </th>
                </tr>
                </thead>
                <tbody className="table__body">
                <SpaciousTableRow className="table__row">
                    <td className="table__cell">
                        Verify if questionnaire is already installed
                    </td>
                    <td className="table__cell">
                        {getIsVerifyIsInstalled()}
                    </td>
                </SpaciousTableRow>
                <SpaciousTableRow className="table__row">
                    <td className="table__cell">
                        Initialise upload
                    </td>
                    <td className="table__cell">
                        {getIsInitialisingUpload()}
                    </td>
                </SpaciousTableRow>
                <SpaciousTableRow className="table__row">
                    <td className="table__cell">
                        Upload questionnaire file
                    </td>
                    <td className="table__cell">
                        {getIsUploading()}
                    </td>
                </SpaciousTableRow>
                <SpaciousTableRow className="table__row">
                    <td className="table__cell">
                        Verify that questionnaire file has been uploaded
                    </td>
                    <td className="table__cell">
                        {getIsVerifyingUpload()}
                    </td>
                </SpaciousTableRow>
                <SpaciousTableRow className="table__row">
                    <td className="table__cell">
                        Install questionnaire
                    </td>
                    <td className="table__cell">
                        {getIsInstalling()}
                    </td>
                </SpaciousTableRow>
                </tbody>
            </table>

        </>
    );
}

export default DeploymentSummary;

function LoadingIcon() {
    return (
        <StyledSVG className="svg-icon uil-default" style={{}} xmlns="http://www.w3.org/2000/svg"
                   viewBox="0 0 100 100"
                   preserveAspectRatio="xMidYMid">
            <rect x="0" y="0" width="100" height="100" fill="none" className="bk"/>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(0 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(30 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.08333333333333333s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(60 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.16666666666666666s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(90 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.25s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(120 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.3333333333333333s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(150 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.4166666666666667s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(180 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.5s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(210 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.5833333333333334s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(240 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.6666666666666666s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(270 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.75s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(300 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.8333333333333334s'
                         repeatCount='indefinite'/>
            </rect>
            <rect x='46.5' y='40' width='7' height='20' rx='5' ry='5'
                  transform='rotate(330 50 50) translate(0 -30)'>
                <animate attributeName='opacity' from='1' to='0' dur='1s' begin='0.9166666666666666s'
                         repeatCount='indefinite'/>
            </rect>
        </StyledSVG>
    );
}
