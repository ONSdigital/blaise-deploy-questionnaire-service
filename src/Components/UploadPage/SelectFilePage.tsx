import {Link} from "react-router-dom";
import {ONSPanel} from "../ONSDesignSystem/ONSPanel";
import {ONSUpload} from "../ONSDesignSystem/ONSUpload";
import {ONSButton} from "../ONSDesignSystem/ONSButton";
import React, {ReactElement} from "react";

interface SelectFilePageProps {
    BeginUploadProcess: any
    setFile: any
    loading: boolean
    panel: string
}

function SelectFilePage(props: SelectFilePageProps): ReactElement {
    const {loading, panel, setFile} = props;

    const handleFileChange = (selectorFiles: FileList | null) => {
        console.log(selectorFiles);
        if (selectorFiles !== null) {
            setFile(selectorFiles);
        }
    };

    return (
        <>
            <Link to="/">
                Previous
            </Link>
            <h1 className="u-mt-s">
                Deploy a questionnaire file
            </h1>

            {panel !== "" && <ONSPanel status="error">{panel}</ONSPanel>}

            <ONSPanel>
                <p>
                    When a questionnaire file is selected and you continue to deploy this questionnaire file, <b>this
                    may take a few minutes</b>.
                    <br/>
                    <br/>
                    Given this, <b>do not navigate away</b> from this page during this process. You will be
                    re-directed
                    when there is an update regarding the deploy of the questionnaire.
                </p>
            </ONSPanel>

            <ONSUpload label="Select survey package"
                       description="File type accepted is .bpkg"
                       fileName="Package"
                       fileID="survey-selector"
                       accept=".bpkg"
                       onChange={(e) => handleFileChange(e.target.files)}
                       disabled={loading}/>

            <ONSButton label="Continue"
                       id="continue-deploy-button"
                       primary={true}
                       onClick={() => props.BeginUploadProcess()}
                       loading={loading}/>
        </>
    );
}

export default SelectFilePage;
