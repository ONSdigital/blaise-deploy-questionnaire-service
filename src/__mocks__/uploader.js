export default function () {
    // return {
    //     options: jest.fn().mockReturnThis(),
    //     continue: jest.fn().mockReturnThis(),
    //     onProgress: jest.fn().mockReturnThis(),
    //     end: jest.fn().mockReturnValue({error: null, data: null}),
    //     abort: jest.fn().mockReturnThis(),
    //     send: this.end(),
    // };

    return {
        options: function (options) {
            return this;
        },
        send: function (file) {
            return this;
        },
        continue: function () {
            return;
        },
        onProgress: function (callback) {
            return this;
        },
        end: function (callback) {
            callback();
        },
        abort: function () {
            return;
        }
    };
}
