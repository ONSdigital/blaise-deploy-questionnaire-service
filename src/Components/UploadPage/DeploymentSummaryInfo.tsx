import React, {ReactElement} from "react";
import {ONSPanel} from "blaise-design-system-react-components";
import {step_status} from "./DeploymentProgress";

interface Props {
    status: string
    instrumentName: string
}

function DeploymentSummaryInfo({status, instrumentName}: Props): ReactElement {
    return (
        (status === "" ?
                <ONSPanel status="success" bigIcon={true}>
                    <h1>
                        Questionnaire
                        file <em>{instrumentName}</em> deployed
                    </h1>
                    <p>
                        The questionnaire file has been successfully deployed and will be displayed within the
                        table
                        of
                        questionnaires.
                    </p>
                </ONSPanel>
                :
                <>
                    <h1>
                        Questionnaire
                        file <em>{instrumentName}</em> deploy failed
                    </h1>
                    <ONSPanel status="error">
                        <p>
                            <b>File deploy failed</b>

                            <br/>
                            <br/>
                            Questionnaire {instrumentName} has failed to deploy. When reporting the issue to
                            Service
                            Desk provide the questionnaire name, time and date of failure.
                        </p>
                        <p>
                            Reason: {status}
                        </p>
                    </ONSPanel>
                </>
        )
    );
}

export default DeploymentSummaryInfo;
