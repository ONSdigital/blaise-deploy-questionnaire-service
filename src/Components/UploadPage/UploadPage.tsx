import React, {ReactElement, useState} from "react";
import {Link, Redirect, Route, Switch, useHistory, useRouteMatch} from "react-router-dom";
import uploader from "../../uploader";
import SelectFilePage from "./SelectFilePage";
import AlreadyExists from "./AlreadyExists";

interface Progress {
    loaded: number
    total: number
}

function UploadPage(): ReactElement {
    const [redirect, setRedirect] = useState<boolean>(false);
    const [loading, setLoading] = useState<boolean>(false);
    const [file, setFile] = useState<FileList>();
    const [fileName, setFileName] = useState<string>("");
    const [instrumentName, setInstrumentName] = useState<string>("");
    const [panel, setPanel] = useState<string>("");
    const [uploadPercentage, setUploadPercentage] = useState<number>(0);
    const [uploadStatus, setUploadStatus] = useState<string>("");
    const timeout = (process.env.NODE_ENV === "test" ? 0 : 3000);

    const {path, url} = useRouteMatch();
    const history = useHistory();

    async function BeginUploadProcess() {
        if (file === undefined) {
            setPanel("You must select a file");
            return;
        }
        if (file.length !== 1) {
            setPanel("Invalid file");
        }
        setLoading(true);
        setFileName(file[0].name);
        setInstrumentName(file[0].name.replace(/\.[a-zA-Z]*$/, ""));

        const alreadyExists = await checkSurveyAlreadyExists(file[0].name.replace(/\.[a-zA-Z]*$/, ""));

        if (alreadyExists) {
            setLoading(false);
            history.push(`${path}/survey-exists`);
        } else {
            await UploadFile();
        }
    }

    async function UploadFile() {
        if (file === undefined) {
            return;
        }
        console.log("Start uploading the file");
        setLoading(true);
        setPanel("");
        uploader()
            .onProgress(({loaded, total}: Progress) => {
                const percent = Math.round(loaded / total * 100 * 100) / 100;
                console.log(`File upload ${percent}% ${loaded} / ${total}`);
                setUploadPercentage(percent);
            })
            .options({
                chunkSize: 5 * 1024 * 1024,
                threadsQuantity: 5
            })
            .send(file[0])
            .end((error: Error) => {
                if (error) {
                    console.log("Failed to upload file, error: ", error);
                    setLoading(false);
                    setUploadStatus("Failed to upload file");
                    setRedirect(true);
                    return;
                }
                console.log("File upload complete");
                setTimeout(function () {
                    checkFileInBucket(file[0].name.replace(/ /g, "_"));
                }, timeout);
            });
    }

    function checkSurveyAlreadyExists(instrumentName: string) {
        console.log("Validating if survey already exists");
        return new Promise((resolve: any, reject: any) => {
            fetch(`/api/instruments/${instrumentName}/exists`)
                .then((r: Response) => {
                    if (r.status !== 200) {
                        throw r.status + " - " + r.statusText;
                    }
                    r.json()
                        .then((json) => {
                            if (json === true) {
                                console.log(`${instrumentName} already installed`);
                                resolve(true);
                            } else {
                                console.log(`${instrumentName} not found `);
                                resolve(false);
                            }
                        })
                        .catch((error) => {
                            console.error("Failed to validate if questionnaire already exists, error: " + error);
                            throw error;
                        });
                })
                .catch(async (error) => {
                    console.error("Failed to validate if questionnaire already exists, error: " + error);
                    await setUploadStatus("Failed to validate if questionnaire already exists");
                    setRedirect(true);
                });
        });
    }


    function checkFileInBucket(filename: string) {
        console.log("Validating file is in the Bucket");
        fetch(`/bucket?filename=${filename}`)
            .then((r: Response) => {
                if (r.status !== 200) {
                    throw r.status + " - " + r.statusText;
                }
                r.json()
                    .then((json) => {
                        if (json.name === filename) {
                            console.log(`File ${filename} successfully uploaded to bucket`);
                            sendInstallRequest(filename);
                        } else {
                            throw "Filename returned does not match sent file";
                        }
                    })
                    .catch((error) => {
                        console.error("Failed to validate if file is in bucket, error: " + error);
                    })
                    .catch((error) => {
                            console.error("Failed to validate if file is in bucket, error: " + error);
                        }
                    );
            })
            .catch(async (error) => {
                console.error("Failed to validate if file is in bucket, error: " + error);
                await setUploadStatus("Failed to validate if file has been uploaded");
                setRedirect(true);
            });
    }

    function sendInstallRequest(filename: string) {
        console.log("Sending request to start install");
        fetch(`/api/install?filename=${filename}`)
            .then((r: Response) => {
                console.log(r);
                if (r.status !== 201) {
                    throw r.status + " - " + r.statusText;
                }
                r.json()
                    .then((json) => {
                        console.log(json);
                        console.log(`Questionnaire ${filename} successfully installed`);
                    })
                    .catch((error) => {
                        console.error("Failed to validate if questionnaire is installed, error: " + error);
                        setLoading(false);
                    });
            })
            .catch((error) => {
                setUploadStatus("Failed to install questionnaire");
                console.error("Failed to install questionnaire, error: " + error);
            }).finally(() => setRedirect(true));
    }

    return (
        <>
            {
                redirect && <Redirect
                    to={{
                        pathname: "/UploadSummary",
                        state: {questionnaireName: instrumentName, status: uploadStatus}
                    }}/>
            }

            <Switch>
                <Route exact path={path}>
                    <SelectFilePage BeginUploadProcess={BeginUploadProcess}
                                    setFile={setFile}
                                    loading={loading}
                                    panel={panel}
                                    uploadPercentage={uploadPercentage}/>
                </Route>
                <Route path={`${path}/survey-exists`}>
                    <AlreadyExists instrumentName={instrumentName}
                                    UploadFile={UploadFile}
                                    loading={loading}/>
                </Route>
            </Switch>

        </>
    );
}

export default UploadPage;
