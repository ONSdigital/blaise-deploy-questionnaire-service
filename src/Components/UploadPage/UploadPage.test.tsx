import React from "react";
import Enzyme from "enzyme";
import {render, waitFor, fireEvent, cleanup, screen} from "@testing-library/react";
import Adapter from "enzyme-adapter-react-16";
import "@testing-library/jest-dom";
import flushPromises from "../../tests/utils";
import {act} from "react-dom/test-utils";
import {createMemoryHistory} from "history";
import {Router} from "react-router";
import {instrumentList} from "../../features/step_definitions/API_Mock_Objects";
import UploadPage from "./UploadPage";
import uploader from "../../uploader";
import navigateToDeployPageAndSelectFile from "../../features/step_definitions/functions";

jest.mock("../../uploader");

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
uploader.__setMockStatus(false);
// Mock the Uploader.js module
// jest.mock("../../uploader");

function mock_server_request(returnedStatus: number, returnedJSON: any) {
    global.fetch = jest.fn(() =>
        Promise.resolve({
            status: returnedStatus,
            json: () => Promise.resolve(returnedJSON),
        })
    );
}

describe("Upload Page", () => {
    Enzyme.configure({adapter: new Adapter()});

    beforeAll(() => {
        mock_server_request(200, instrumentList);
    });

    it("select file page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <UploadPage/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <UploadPage/>
            </Router>
        );

        await waitFor(() => {
            expect(getByText(/Deploy a questionnaire file/i)).toBeDefined();
            expect(getByText(/Select survey package/i)).toBeDefined();
            expect(queryByText(/Table of questionnaires/i)).not.toBeInTheDocument();
        });
    });

    it("should display a message if you dont select a file", async () => {
        const history = createMemoryHistory();
        const {getByText, getByTestId} = render(
            <Router history={history}>
                <UploadPage/>
            </Router>
        );

        await fireEvent.click(getByTestId("button"));

        await waitFor(() => {
            expect(getByText(/You must select a file/i)).toBeDefined();
        });
    });

    it("should display a message if select a file that is not .zip or .bpkg", async () => {
        const history = createMemoryHistory();
        const {getByText, getByLabelText, getByTestId} = render(
            <Router history={history}>
                <UploadPage/>
            </Router>
        );

        const inputEl = getByLabelText(/Select survey package/i);

        const file = new File(["(⌐□_□)"], "OPN2004A.pdf", {
            type: "pdf"
        });

        Object.defineProperty(inputEl, "files", {
            value: [file]
        });

        fireEvent.change(inputEl);

        await fireEvent.click(getByTestId("button"));

        await waitFor(() => {
            expect(getByText(/File must be a .zip or .bpkg/i)).toBeDefined();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});


describe("Given the file fails to upload", () => {
    Enzyme.configure({adapter: new Adapter()});

    beforeAll(() => {
        mock_server_request(200, instrumentList);
    });

    it("it should redirect to the summary page with an error", async () => {
        await navigateToDeployPageAndSelectFile();

        await fireEvent.click(screen.getByTestId("button"));

        await waitFor(() => {
            expect(screen.getByText("File deploy failed")).toBeDefined();
            expect(screen.getByText(/Failed to upload file/i)).toBeDefined();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
