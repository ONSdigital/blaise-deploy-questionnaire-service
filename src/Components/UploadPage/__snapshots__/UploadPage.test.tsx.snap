// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Page select file page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/"
      >
        Previous
      </a>
      <h1
        class="u-mt-s"
      >
        Deploy a questionnaire file
      </h1>
      <div
        class="panel panel--info panel--no-title  u-mt-m"
      >
        <div
          class="panel__body"
        >
          <p>
            When a questionnaire file is selected and you continue to deploy this questionnaire file, 
            <b>
              this may take a few minutes
            </b>
            .
            <br />
            <br />
            Given this, 
            <b>
              do not navigate away
            </b>
             from this page during this process. You will be re-directed when there is an update regarding the deploy of the questionnaire.
          </p>
        </div>
      </div>
      <div
        class="field"
      >
        <p
          class="field"
        >
          <label
            class="label"
            for="survey-selector"
          >
            Select survey package
            <br />
            <span
              class="label__description"
            >
              File type accepted are .bpkg and .zip only
            </span>
          </label>
          <input
            accept=".zip,.bpkg"
            class="input input--text input-type__input input--upload"
            id="survey-selector"
            name="Package"
            style="position: static;"
            type="file"
          />
        </p>
      </div>
      <button
        class="btn "
        data-testid="button"
        id="continue-deploy-button"
        type="button"
      >
        <span
          class="btn__inner"
        >
          Continue
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <a
      href="/"
    >
      Previous
    </a>
    <h1
      class="u-mt-s"
    >
      Deploy a questionnaire file
    </h1>
    <div
      class="panel panel--info panel--no-title  u-mt-m"
    >
      <div
        class="panel__body"
      >
        <p>
          When a questionnaire file is selected and you continue to deploy this questionnaire file, 
          <b>
            this may take a few minutes
          </b>
          .
          <br />
          <br />
          Given this, 
          <b>
            do not navigate away
          </b>
           from this page during this process. You will be re-directed when there is an update regarding the deploy of the questionnaire.
        </p>
      </div>
    </div>
    <div
      class="field"
    >
      <p
        class="field"
      >
        <label
          class="label"
          for="survey-selector"
        >
          Select survey package
          <br />
          <span
            class="label__description"
          >
            File type accepted are .bpkg and .zip only
          </span>
        </label>
        <input
          accept=".zip,.bpkg"
          class="input input--text input-type__input input--upload"
          id="survey-selector"
          name="Package"
          style="position: static;"
          type="file"
        />
      </p>
    </div>
    <button
      class="btn "
      data-testid="button"
      id="continue-deploy-button"
      type="button"
    >
      <span
        class="btn__inner"
      >
        Continue
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
